configs/compat_shim.go

```diff
@@ -107,3 +107,58 @@ func shimIsIgnoreChangesStar(expr hcl.Expression) bool {
 	}
 	return val.AsString() == "*"
 }
+
+// warnForDeprecatedInterpolations returns warning diagnostics if the given
+// body can be proven to contain attributes whose expressions are native
+// syntax expressions consisting entirely of a single template interpolation,
+// which is a deprecated way to include a non-literal value in configuration.
+//
+// This is a best-effort sort of thing which relies on the physical HCL native
+// syntax AST, so it might not catch everything. The main goal is to catch the
+// "obvious" cases in order to help spread awareness that this old form is
+// deprecated, when folks copy it from older examples they've found on the
+// internet that were written for Terraform 0.11 or earlier.
+func warnForDeprecatedInterpolationsInBody(body hcl.Body) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	nativeBody, ok := body.(*hclsyntax.Body)
+	if !ok {
+		// If it's not native syntax then we've nothing to do here.
+		return diags
+	}
+
+	for _, attr := range nativeBody.Attributes {
+		moreDiags := warnForDeprecatedInterpolationsInExpr(attr.Expr)
+		diags = append(diags, moreDiags...)
+	}
+
+	for _, block := range nativeBody.Blocks {
+		// We'll also go hunting in nested blocks
+		moreDiags := warnForDeprecatedInterpolationsInBody(block.Body)
+		diags = append(diags, moreDiags...)
+	}
+
+	return diags
+}
+
+func warnForDeprecatedInterpolationsInExpr(expr hcl.Expression) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	if _, ok := expr.(*hclsyntax.TemplateWrapExpr); !ok {
+		// We're only interested in TemplateWrapExpr, because that's how
+		// the HCL native syntax parser represents the case of a template
+		// that consists entirely of a single interpolation expression, which
+		// is therefore subject to the special case of passing through the
+		// inner value without conversion to string.
+		return diags
+	}
+
+	diags = append(diags, &hcl.Diagnostic{
+		Severity: hcl.DiagWarning,
+		Summary:  "Interpolation-only expressions are deprecated",
+		Detail:   "Terraform 0.11 and earlier required all non-constant expressions to be provided via interpolation syntax, but this pattern is now deprecated. To silence this warning, remove the \"${ sequence from the start and the }\" sequence from the end of this expression, leaving just the inner expression.\n\nTemplate interpolation syntax can still be used to construct strings from expressions when the template includes multiple interpolation sequences or a mixture of literal strings and interpolations. This deprecation applies only to templates that consist entirely of a single interpolation sequence.",
```

Comment:
Nitpick: The word "can" in the third sentence sounds like it's optional. What do you think about something like this change: 

Orig: Template interpolation syntax can still be used to ...
Suggestion: Template interpolation syntax is (or "is still", "is required to", etc) used to  ...

Reply:
Good call! I'll change it to "is still used".